sdk/communication/Azure.Communication.Common/src/EntraTokenCredential.cs

```diff
@@ -0,0 +1,191 @@
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token for the task.</param>
+        /// <returns> Contains the access token.</returns>
+        public AccessToken GetToken(CancellationToken cancellationToken)
+        {
+            return GetTokenAsync(cancellationToken).Result;
+        }
+
+        /// <summary>
+        /// Gets an <see cref="AccessToken"/>.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token for the task.</param>
+        /// <returns>
+        /// A task that represents the asynchronous get token operation. The value of its <see cref="ValueTask{AccessToken}.Result"/> property contains the access token.
+        /// </returns>
+        public ValueTask<AccessToken> GetTokenAsync(CancellationToken cancellationToken)
+        {
+            return GetValueAsync(cancellationToken);
+        }
+
+        private async ValueTask<AccessToken> GetValueAsync(CancellationToken cancellationToken)
+        {
+            if (IsTokenValid(_currentToken))
+                return _currentToken;
+
+            var shouldThisThreadExchange = false;
+            lock (_syncLock)
+            {
+                while (!IsTokenValid(_currentToken))
+                {
+                    if (_someThreadIsExchanging)
+                    {
+                        if (IsTokenValid(_currentToken))
+                            return _currentToken;
+
+                        WaitForInProgressThreadFinish();
+                    }
+                    else
+                    {
+                        shouldThisThreadExchange = true;
+                        _someThreadIsExchanging = true;
+                        break;
+                    }
+                }
+            }
+
+            if (shouldThisThreadExchange)
+            {
+                try
+                {
+                    AccessToken result = await ExchangeEntraToken(cancellationToken).ConfigureAwait(false);
+                    lock (_syncLock)
+                    {
+                        _currentToken = result;
+                        Thread.MemoryBarrier();
+                        _someThreadIsExchanging = false;
+                        NotifyTokenExchangeDone();
+                    }
+                }
+                catch
+                {
+                    lock (_syncLock)
+                    {
+                        _someThreadIsExchanging = false;
+                        NotifyTokenExchangeDone();
+                    }
+                    throw;
+                }
+            }
+
+            return _currentToken;
+
+            void WaitForInProgressThreadFinish()
+                => Monitor.Wait(_syncLock);
+
+            void NotifyTokenExchangeDone()
+                => Monitor.PulseAll(_syncLock);
+        }
+
+        private bool IsTokenValid(AccessToken? token)
+            => token != null && DateTimeOffset.UtcNow < token?.ExpiresOn;
```

Comment:
In my JS PR, if I remember correctly, I wanted to ensure that having a new Entra token will always get a new ACS token. That way the ACS token just follows whatever refresh logic the MSAL tokenCredential has internally.

Reply:
added the logic for cashing entra token